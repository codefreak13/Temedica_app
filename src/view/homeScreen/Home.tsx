import React, { useState, useCallback } from 'react';
import { FlatList, Keyboard } from 'react-native';
import debounce from 'lodash.debounce';
import styled from 'styled-components/native'
import Ionicons from 'react-native-vector-icons/Ionicons'

import dataSet from '../../data/dataset.json'
import ListItem, { Text, View } from './HomeItem';
import { colors } from './../../colors'
import logo from '../../assets/images/logo.png'

const Touchable = styled.TouchableOpacity``

const Wrapper = styled(View)`
flex:1;
margin: 15px
`
const InputWrapper = styled(View)`
flex-direction: row;
align-items: center;
justify-content: space-between;
border-color: ${colors.BLACK};
border-width: 0.5px;
border-radius: 10px;
padding-horizontal: 5px
`
const Input = styled.TextInput`
padding-horizontal:10px;
font-size: 15px;
color: ${colors.BLACK};
flex:1;
`
const Logo = styled.Image`
width: 70px;
height:70px;
align-self: center;
`
const ShowText = styled(Text)`
margin-vertical: 10px
`
const EmptyText = styled(Text)`
margin-vertical: 150px;
align-self: center;
font-size: 20px
`
const ListWrapper = styled(View)`
flex:1;
`
type HomeProps = {
    id: string,
    name: string,
    diseases: Array<string>,
    description: string,
    released: string
}

const Home = () => {
    const [value, setvalue] = useState<string>('');
    const [data, setdata] = useState<HomeProps[]>([]);

    const fetchData = async (searchText: string) => {
        const matcher = new RegExp(searchText, 'ig');

        const selectedData = dataSet?.drugs?.filter((item) => {
            const { name, diseases } = item;

            return (matcher.test(name) && searchText !== '') || diseases.filter((disease) => matcher.test(disease)).length
        })

        Keyboard.dismiss()
        setdata(selectedData)
    };

    const debouncedSave = useCallback(
        debounce((nextValue: string) => fetchData(nextValue), 1000),
        []
    );

    const onChangeText = (text: string) => {
        setvalue(text);
        debouncedSave(text);
    };

    const clearText = () => {
        setvalue('')
        setdata([])
    }

    return (
        <Wrapper>
            <Logo source={logo} />
            <ShowText>Search</ShowText>
            <InputWrapper>
                <Input
                    placeholder='You can search by drug name or by disease'
                    onChangeText={text => onChangeText(text)}
                    value={value}
                />
                {!!value && <Touchable onPress={() => { clearText() }}>
                    <Ionicons name="close" size={20} />
                </Touchable>}
            </InputWrapper>

            <ListWrapper>
                {data.length > 0 && (
                    <ShowText>Showing {data.length} result(s)</ShowText>)}
                <FlatList
                    data={data}
                    keyExtractor={(item: HomeProps) => item.id}
                    renderItem={({ item }) => <ListItem {...item} />}
                    showsVerticalScrollIndicator={false}
                    ListEmptyComponent={<EmptyText>No Drugs</EmptyText>}
                />
            </ListWrapper>
        </Wrapper>
    );
};

export default Home;
