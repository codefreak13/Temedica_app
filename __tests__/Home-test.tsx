import 'react-native'
import React from 'react';
import Home from '../src/view/homeScreen/Home';
import { render } from '@testing-library/react-native';

describe('The home component renders correctly', () =>
it('and placeholder text is correct', () => {
    const { getByPlaceholderText, toJSON } = render(<Home/>);

    const placeholderText = getByPlaceholderText('You can search by drug name or by disease');

    expect(placeholderText).toBeTruthy()

    expect(toJSON()).toMatchSnapshot()
})
)