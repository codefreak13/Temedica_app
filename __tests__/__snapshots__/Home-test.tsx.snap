// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The home component renders correctly and placeholder text is correct 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "marginHorizontal": 15,
        "marginTop": 70,
      },
    ]
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../src/assets/images/logo.png",
      }
    }
    style={
      Array [
        Object {
          "alignSelf": "center",
          "height": 70,
          "width": 70,
        },
      ]
    }
    testID="logo"
  />
  <Text
    style={
      Array [
        Object {
          "color": "#000",
          "marginVertical": 10,
        },
      ]
    }
    testID="search"
  >
    Search
  </Text>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderColor": "#000",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 5,
        },
      ]
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="You can search by drug name or by disease"
      style={
        Array [
          Object {
            "color": "#000",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 15,
            "minHeight": 45,
            "paddingHorizontal": 10,
          },
        ]
      }
      testID="inputText"
      value=""
    />
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <RCTScrollView
      ListEmptyComponent={
        <Styled(Styled(Text))
          testID="emptyText"
        >
          Please Input Search Item
        </Styled(Styled(Text))>
      }
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      testID="listId"
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <Text
          onLayout={[Function]}
          style={
            Array [
              Object {
                "alignSelf": "center",
                "color": "#000",
                "fontSize": 17,
                "marginVertical": 200,
              },
            ]
          }
          testID="emptyText"
        >
          Please Input Search Item
        </Text>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
